crates/accelerate/src/permutation.rs

```diff
@@ -0,0 +1,122 @@
+    let mut seen: Vec<bool> = vec![false; n];
+
+    for &x in pattern {
+        if x < 0 {
+            return Err(PyValueError::new_err(
+                "Invalid permutation: input contains a negative number.",
+            ));
+        }
+
+        if x as usize >= n {
+            return Err(PyValueError::new_err(format!(
+                "Invalid permutation: input has length {} and contains {}.",
+                n, x
+            )));
+        }
+
+        if seen[x as usize] {
+            return Err(PyValueError::new_err(format!(
+                "Invalid permutation: input contains {} more than once.",
+                x
+            )));
+        }
+
+        seen[x as usize] = true;
+    }
+
+    Ok(())
+}
+
+fn invert(pattern: &ArrayView1<i64>) -> Array1<usize> {
+    let mut inverse: Array1<usize> = Array1::zeros(pattern.len());
+    pattern.iter().enumerate().for_each(|(ii, &jj)| {
+        inverse[jj as usize] = ii;
+    });
+    inverse
+}
+
+fn get_ordered_swap(pattern: &ArrayView1<i64>) -> Vec<(i64, i64)> {
+    let mut permutation: Vec<usize> = pattern.iter().map(|&x| x as usize).collect();
+    let mut index_map = invert(pattern);
+
+    let n = permutation.len();
+    let mut swaps: Vec<(i64, i64)> = Vec::with_capacity(n);
+    for ii in 0..n {
+        let val = permutation[ii];
+        if val == ii {
+            continue;
+        }
+        let jj = index_map[ii];
+        swaps.push((ii as i64, jj as i64));
+        (permutation[ii], permutation[jj]) = (permutation[jj], permutation[ii]);
+        index_map[val] = jj;
+        index_map[ii] = ii;
+    }
+
+    swaps[..].reverse();
+    swaps
+}
+
+/// Finds inverse of a permutation pattern.
+#[pyfunction]
+#[pyo3(signature = (pattern))]
+fn _inverse_pattern(py: Python, pattern: PyArrayLike1<i64, AllowTypeChange>) -> PyResult<PyObject> {
+    let view = pattern.as_array();
+    validate_permutation(&view)?;
+    let inverse_i64: Vec<i64> = invert(&view).iter().map(|&x| x as i64).collect();
+    Ok(inverse_i64.to_object(py))
+}
+
+/// Sorts the input permutation by iterating through the permutation list
+/// and putting each element to its correct position via a SWAP (if it's not
+/// at the correct position already). If ``n`` is the length of the input
+/// permutation, this requires at most ``n`` SWAPs.
+///
+/// More precisely, if the input permutation is a cycle of length ``m``,
+/// then this creates a quantum circuit with ``m-1`` SWAPs (and of depth ``m-1``);
+/// if the input  permutation consists of several disjoint cycles, then each cycle
+/// is essentially treated independently.
+#[pyfunction]
+#[pyo3(signature = (permutation_in))]
+fn _get_ordered_swap(
+    py: Python,
+    permutation_in: PyArrayLike1<i64, AllowTypeChange>,
+) -> PyResult<PyObject> {
+    let view = permutation_in.as_array();
+    validate_permutation(&view)?;
+    Ok(get_ordered_swap(&view).to_object(py))
+}
+
+#[pymodule]
+pub fn permutation_utils(m: &Bound<PyModule>) -> PyResult<()> {
```

Comment:
why do you need two `[pymodule]`s here? I see that in other cases only one was used.

Reply:
It's not really necessary. I copied the structure from the Python `synthesis` module:

```
qiskit
    |-- synthesis
        |-- permutation
            | -- permutation_full
            | -- permutation_lnn
            | -- permutation_reverse_lnn
            | -- permutation_utils
```

since I was thinking the idea is to eventually move all of it to Rust.

Reply:
you are right that the idea is to eventually move all of the synthesis code to rust,
however, it seems that currently the structure is flat, see e.g.
https://github.com/Qiskit/qiskit/blob/6e3bc2c292d697eb180f75e9943f54839c553038/qiskit/__init__.py#L83
https://github.com/Qiskit/qiskit/blob/6e3bc2c292d697eb180f75e9943f54839c553038/qiskit/__init__.py#L70

Reply:
Got it, I made another commit to flatten this as well.